File: app/src/main/java/uk/co/transferx/app/recoverpass/presenter/RecoverPasswordPresenter.java
package uk.co.transferx.app.recoverpass.presenter;

import javax.inject.Inject;
import javax.net.ssl.HttpsURLConnection;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import uk.co.transferx.app.BasePresenter;
import uk.co.transferx.app.UI;
import uk.co.transferx.app.api.SignInOutApi;
import uk.co.transferx.app.tokenmanager.TokenManager;
import uk.co.transferx.app.util.Util;

/**
 * Created by sergey on 15/03/2018.
 */

public class RecoverPasswordPresenter extends BasePresenter<RecoverPasswordPresenter.RecoverPasswordUI> {

    private final TokenManager tokenManager;
    private String validEmail;
    private final SignInOutApi signInOutApi;
    private Disposable disposable;

    @Inject
    public RecoverPasswordPresenter(final TokenManager tokenManager, final SignInOutApi signInOutApi) {
        this.tokenManager = tokenManager;
        this.signInOutApi = signInOutApi;
    }

    @Override
    public void detachUI() {
        super.detachUI();
        if (disposable != null) {
            disposable.dispose();
        }
    }

    public void validateEmail(String email) {
        if (!Util.validateEmail(email)) {
            ui.lockButton();
            return;
        }
        validEmail = email;
        ui.unlockButton();
    }

    public void sendEmail() {
        disposable = tokenManager.getToken()
                .flatMap(token -> signInOutApi.forgotEmail(token.getAccessToken(), validEmail))
                .subscribeOn(Schedulers.io())
                .subscribeOn(AndroidSchedulers.mainThread())
                .subscribe(resp -> {
                    if (ui != null && resp.code() == HttpsURLConnection.HTTP_OK) {
                        ui.goBackToMain();
                        return;
                    }
                    handleError(new Throwable(resp.message()));
                }, this::handleError);
    }

    public void goBack() {
        ui.goBackToMain();
    }

    public void handleError(Throwable throwable){
        if(ui != null){
            ui.error();
        }
    }

    public interface RecoverPasswordUI extends UI {

        void unlockButton();

        void successGoBack();

        void lockButton();

        void goBackToMain();

        void error();

    }
}

File: app/src/main/java/uk/co/transferx/app/ui/recoverpass/presenter/RecoverPasswordPresenter.kt
package uk.co.transferx.app.ui.recoverpass.presenter

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.disposables.Disposables
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody
import retrofit2.Response
import uk.co.transferx.app.data.pojo.ForgotEmail
import uk.co.transferx.app.data.remote.SignInOutApi
import uk.co.transferx.app.data.repository.tokenmanager.TokenManager
import uk.co.transferx.app.ui.base.BasePresenter
import uk.co.transferx.app.ui.recoverpass.RecoverPasswordContract
import uk.co.transferx.app.util.Util
import javax.inject.Inject
import javax.net.ssl.HttpsURLConnection

/**
 * Created by sergey on 15/03/2018.
 */

class RecoverPasswordPresenter @Inject constructor
(private val tokenManager: TokenManager, private val signInOutApi: SignInOutApi)
    : BasePresenter<RecoverPasswordContract.View>(), RecoverPasswordContract.Presenter {
    private var validEmail: String? = null
    private var disposable: Disposable? = Disposables.disposed()

    override fun detachUI() {
        super.detachUI()
        disposable?.dispose()
    }

    override
    fun validateEmail(email: String) {
        if (!Util.validateEmail(email)) {
            this.ui?.lockButton()
        } else {
            validEmail = email
            ui.unlockButton()
        }
    }

    override
    fun sendEmail() {
        disposable = tokenManager.token
                .flatMap<Response<ResponseBody>> { (accessToken) -> signInOutApi.forgotEmail(accessToken, ForgotEmail(validEmail!!)) }
                .subscribeOn(Schedulers.io())
                .subscribeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it.code() == HttpsURLConnection.HTTP_OK) this.ui?.goToSuccessScreen()
                    else handleError(Throwable(it.message()))
                }, { this.handleError(it) })
    }

    override
    fun goBack() {
        this.ui?.goBackToMain()
    }

    override
    fun handleError(throwable: Throwable) {
        this.ui?.error()
    }

}
